server/src/main/java/org/apache/cloudstack/snapshot/SnapshotHelper.java

```diff
@@ -0,0 +1,255 @@
+            return DataStoreRole.Image;
+        }
+
+        Map<String, String> mapCapabilities = dataStore.getDriver().getCapabilities();
+
+        if (MapUtils.isNotEmpty(mapCapabilities) && BooleanUtils.toBoolean(mapCapabilities.get(DataStoreCapabilities.STORAGE_SYSTEM_SNAPSHOT.toString()))) {
+            return DataStoreRole.Primary;
+        }
+
+        return DataStoreRole.Image;
+    }
+
+    /**
+     * Verifies if it is a KVM volume that has snapshots only in primary storage.
+     * @throws CloudRuntimeException If it is a KVM volume and has at least one snapshot only in primary storage.
+     */
+    public void isKvmVolumeSnapshotsOnlyInPrimaryStorage(VolumeVO volumeVo, HypervisorType hypervisorType) throws CloudRuntimeException {
+        if (HypervisorType.KVM != hypervisorType) {
+            logger.trace(String.format("The %s hypervisor [%s] is not KVM, therefore we will not check if the snapshots are only in primary storage.", volumeVo, hypervisorType));
+            return;
+        }
+
+        Set<Long> snapshotIdsOnlyInPrimaryStorage = getSnapshotIdsOnlyInPrimaryStorage(volumeVo.getId());
+
+        if (CollectionUtils.isEmpty(snapshotIdsOnlyInPrimaryStorage)) {
+            logger.trace(String.format("%s is a KVM volume and all its snapshots exists in the secondary storage, therefore this volume is able for migration.", volumeVo));
+            return;
+        }
+
+        throwCloudRuntimeExceptionOfSnapshotsOnlyInPrimaryStorage(volumeVo, snapshotIdsOnlyInPrimaryStorage);
+    }
+
+    /**
+     * Throws a CloudRuntimeException with the volume and the snapshots only in primary storage.
+     */
+    protected void throwCloudRuntimeExceptionOfSnapshotsOnlyInPrimaryStorage(VolumeVO volumeVo, Set<Long> snapshotIdsOnlyInPrimaryStorage) throws CloudRuntimeException {
+        List<SnapshotVO> snapshots = snapshotDao.listByIds(snapshotIdsOnlyInPrimaryStorage.toArray());
+
+        String message = String.format("%s is a KVM volume and has snapshots only in primary storage. Snapshots [%s].%s", volumeVo,
+                snapshots.stream().map(snapshot -> new ToStringBuilder(snapshot, ToStringStyle.JSON_STYLE).append("uuid", snapshot.getUuid()).append("name", snapshot.getName())
+                        .build()).collect(Collectors.joining(", ")), backupSnapshotAfterTakingSnapshot ? "" : " Consider excluding them to migrate the volume to another storage.");
+
+        logger.error(message);
+        throw new CloudRuntimeException(message);
+    }
+
+    /**
+     * Retrieves the ids of the ready snapshots of the volume that only exists in primary storage.
+     * @param volumeId volume id to retrieve the snapshots.
+     * @return The ids of the ready snapshots of the volume that only exists in primary storage
+     */
+    protected Set<Long> getSnapshotIdsOnlyInPrimaryStorage(long volumeId) {
+        List<SnapshotDataStoreVO> snapshotsReferences = snapshotDataStoreDao.listReadyByVolumeId(volumeId);
```

Comment:
You can directly list all ready snapshots by volume id and data store role 
```suggestion
        List<SnapshotDataStoreVO> snapshotsReferences = snapshotDataStoreDao.listReadyByVolumeId(volumeId, DataStoreRole.Primary);
```